### PR TITLE
Add date-range picker for referrals

### PR DESCRIPTION
**Pull request by kelvin-chappell**
image

### CODE REVIEW

File: public/components/CampaignAnalytics/Analytics/CampaignReferrals.js [hunk 1]
```diff
@@ -5,60 +5,77 @@ import {getCampaignReferrals, setToggleNode} from "../../../actions/CampaignActi
...
   componentWillMount() {
-    this.props.campaignReferralActions.getCampaignReferrals(this.props.campaign.id);
+    this.props.campaignReferralActions.getCampaignReferrals(this.props.campaign.id, this.state.dateRange);
   }
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.campaign.id !== this.props.campaign.id) {
-      this.props.campaignReferralActions.getCampaignReferrals(nextProps.campaign.id);
+  componentWillUpdate(nextProps, nextState) {
+    if (nextProps.campaign.id !== this.props.campaign.id || nextState !== this.state) {
```
Comment (rich-nguyen) [inline public/components/CampaignAnalytics/Analytics/CampaignReferrals.js]: I'm not sure if that test really captures specific-case update criteria. Is this supposed to detect changes in `dateRange`? could it be more specific? also, the params could just be an object, that encapsulates all the state that parameterises the query (the campaign id, date range). It's also a bit confusing having redux and `componentWillUpdate` overlapping each other.

File: public/components/CampaignAnalytics/Analytics/DateRangeEditor.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  handleChange({target}) {
+    this.setState({
+      [target.name]: target.value
+    });
+  }
+
+  static zeroPad(num) {
+    return ('0' + num).slice(-2)
```
Comment (rich-nguyen) [inline public/components/CampaignAnalytics/Analytics/DateRangeEditor.js:27]: I recognise this time-formatting snippet from stackoverflow. a classic!

**Comment by rich-nguyen** [comment]
It looks great, just that one issue about the state. Admittedly, it was mainly me moving the filtering out of the component and into the global redux store that has complicated things. But if the three parameters that influenced the tree were in one place, it would be fairly readable (three parameters being date, campaign id, ordering).

File: public/components/CampaignAnalytics/Analytics/DateRangeEditor.js (hunk 2 above)
Comment (kelvin-chappell) [reply public/components/CampaignAnalytics/Analytics/DateRangeEditor.js:27]: The source of all my knowledge.  😄

File: public/components/CampaignAnalytics/Analytics/CampaignReferrals.js (hunk 1 above)
Comment (kelvin-chappell) [reply public/components/CampaignAnalytics/Analytics/CampaignReferrals.js]: @rich-nguyen Is this clearer?  I don't think the change in properties is actually relevant so it's just a particular change in state that I need to look out for.

File: public/components/CampaignAnalytics/Analytics/CampaignReferrals.js [hunk 3]
```diff
@@ -5,60 +5,77 @@ import {getCampaignReferrals, setToggleNode} from "../../../actions/CampaignActi
...
   componentWillMount() {
-    this.props.campaignReferralActions.getCampaignReferrals(this.props.campaign.id);
+    this.props.campaignReferralActions.getCampaignReferrals(this.props.campaign.id, this.state.dateRange);
   }
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.campaign.id !== this.props.campaign.id) {
-      this.props.campaignReferralActions.getCampaignReferrals(nextProps.campaign.id);
+  componentWillUpdate(nextProps, nextState) {
+    if (nextState.dateRange !== this.state.dateRange) {
```
Comment (rich-nguyen) [inline public/components/CampaignAnalytics/Analytics/CampaignReferrals.js:27]: what happened to the campaign id test?

**Comment by rich-nguyen** [comment]
👍

File: public/components/CampaignAnalytics/Analytics/CampaignReferrals.js (hunk 3 above)
Comment (kelvin-chappell) [reply public/components/CampaignAnalytics/Analytics/CampaignReferrals.js:27]: I don't think we need it, but will reinstate if anything strange happens.

**Pull request merged**